### PR TITLE
Recoverable tasks [version 1]

### PR DESCRIPTION
**Pull request by acierto**
## Hey, I just made a Pull Request!

That's a slice of this [PR](https://github.com/backstage/backstage/pull/21299).
Included the logic to restart the task from the beginning. 
It makes the logic way more simple to create a good basis to build on top of it more smart way of recovering tasks after server crash or redeploy.
At this stage engineer is responsible to make sure that the task is fully idempotent and can run all steps once again.
It means that at this moment you can't use standard actions out-of-the-box, but come up with your own solution

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: docs/features/software-templates/recoverable-tasks.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+And if at this moment some tasks (templates being executed) were in progress, by default, the execution
+will be terminated. Till task janitor close stale tasks for you, you keep seeing the status of these tasks
+as "in progress", but in fact they are not active anymore.
+
+If you would like to recover your tasks automatically, at this version you'll have an option to restart your task.
+
+For that you have to specify in your template a recovery strategy:
+
+```yaml
+apiVersion: scaffolder.backstage.io/v1beta4
```
Comment (benjdlambert) [inline docs/features/software-templates/recoverable-tasks.md:17]: I think that we should just do it in the v1beta3 spec for now, kind of hoping that we can save beta4 for maybe liquidjs and things.

File: docs/features/software-templates/recoverable-tasks.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+kind: Template
+metadata:
+  name: my-recoverable-template
+spec:
+  recovery:
```
Comment (benjdlambert) [inline docs/features/software-templates/recoverable-tasks.md:22]: Think it should be `EXPERIMENTAL_recovery` instead, then anything underneath it is treated as experimental.

Comment (acierto) [reply docs/features/software-templates/recoverable-tasks.md:22]: oh yeah, good catch, missed md file 🙈

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts [hunk 3]
```diff
@@ -451,6 +472,32 @@ export class DatabaseTaskStore implements TaskStore {
     });
   }
 
+  private async reopenTask(options: { taskId: string }): Promise<void> {
+    const { taskId } = options;
+
+    await this.db.transaction(async tx => {
+      const [{ spec }] = await tx<RawDbTaskRow>('tasks')
+        .where({ id: taskId })
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:480]: status = 'stale'?

File: plugins/scaffolder-backend/src/ScaffolderPlugin.ts [hunk 4]
```diff
@@ -75,6 +75,7 @@ export const scaffolderPlugin = createBackendPlugin({
       deps: {
         logger: coreServices.logger,
         config: coreServices.rootConfig,
+        lifecycle: coreServices.lifecycle,
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/ScaffolderPlugin.ts:78]: `coreServices.rootLifecycle`

File: plugins/scaffolder-backend/src/scaffolder/tasks/types.ts [hunk 5]
```diff
@@ -142,6 +143,14 @@ export type TaskStoreCreateTaskOptions = {
   secrets?: TaskSecrets;
 };
 
+/**
+ * The options passed to {@link TaskStore.recoverTasks}
+ * @public
+ */
+export type TaskStoreRecoverTaskOptions = {
+  timeoutS: number;
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/types.ts:151]: Think we should make this `timeout: HumanDuration`.

Looks like this:

`timeout: { seconds: 100 }` in practice

File: plugins/scaffolder-backend/src/service/router.ts [hunk 6]
```diff
@@ -331,7 +335,21 @@ export async function createRouter(
       });
 
   actionsToRegister.forEach(action => actionRegistry.register(action));
-  workers.forEach(worker => worker.start());
+
+  if (options.eventBroker) {
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/service/router.ts:339]: let's drop this for now, and if there's no lifecycle service just start the workers.

File: plugins/scaffolder-common/src/Template.v1beta3.schema.json [hunk 7]
```diff
@@ -139,6 +139,34 @@
                 }
               ]
             },
+            "presentation": {
```
Comment (benjdlambert) [inline plugins/scaffolder-common/src/Template.v1beta3.schema.json:142]: Can drop this presentation stuff

File: plugins/scaffolder-common/src/TemplateEntityV1beta3.ts [hunk 8]
```diff
@@ -73,6 +78,22 @@ export interface TemplateEntityV1beta3 extends Entity {
...
+ */
+export interface TemplateRecoveryV1beta3 extends JsonObject {
+  /**
+   *
+   * none - not recover, let the task be marked as failed
+   * start_over - do recover, start the execution of the task from the first step.
+   *
+   * @public
+   */
+  EXPERIMENTAL_strategy?: 'none' | 'start_over';
```
Comment (benjdlambert) [inline plugins/scaffolder-common/src/TemplateEntityV1beta3.ts:94]:
```suggestion
  EXPERIMENTAL_strategy?: 'none' | 'startOver';
```

**Comment by acierto** [comment]
Created another PR due to DCO issues: https://github.com/backstage/backstage/pull/22236

**Pull request closed without merging**